pykrige/kriging_tools.py

```diff
@@ -246,3 +248,181 @@ def read_asc_grid(filename, footer=0):
         cellsize = (dx, dy)
 
     return grid_array, x, y, cellsize, no_data
+
+
+def write_zmap_grid(x, y, z, filename="output.zmap", no_data=-999.0, coord_sys='<null>'):
+    r"""Writes gridded data to ASCII grid file in zmap format (\*.zmap).
+
+    This is useful for exporting data to a GIS program, or Petrel
+    https://gdal.org/drivers/raster/zmap.html
+
+    Parameters
+    ----------
+    x : array_like, shape (N,) or (N, 1)
+        X-coordinates of grid points at center of cells.
+    y : array_like, shape (M,) or (M, 1)
+        Y-coordinates of grid points at center of cells.
+    z : array_like, shape (M, N)
+        Gridded data values. May be a masked array.
+    filename : string, optional
+        Name of output \*.zmap file. Default name is 'output.zmap'.
+    no_data : float, optional
+        no data value to be used
+    coord_sys : String, optional
+        coordinate sytem description
+    """
+
+    nodes_per_line = 5
+    field_width = 15
+
+    if np.ma.is_masked(z):
+        z = np.array(z.tolist(no_data))
+
+    x = np.squeeze(np.array(x))
+    y = np.squeeze(np.array(y))
+    z = np.squeeze(np.array(z))
+    nx = len(x)
+    ny = len(y)
+
+    dx = abs(x[1] - x[0])
+    dy = abs(y[1] - y[0])
+
+    xllcenter = x[0]
+    yllcenter = y[0]
+
+    hix = xllcenter + (nx - 1) * dx
+    hiy = yllcenter + (ny - 1) * dy
+
+    now = datetime.datetime.now()
+
+    with io.open(filename, "w") as f:
+        f.write("!" + "\n")
+        f.write("!     ZIMS FILE NAME :  " + os.path.basename(filename) + "\n")
+        f.write("!     FORMATTED FILE CREATION DATE: " + now.strftime("%d/%m/%Y") + "\n")
+        f.write("!     FORMATTED FILE CREATION TIME: " + now.strftime("%H:%M:%S") + "\n")
+        f.write("!     COORDINATE REFERENCE SYSTEM: " + coord_sys + "\n")
+        f.write("!" + "\n")
+        f.write("@Grid HEADER, GRID, " + str(nodes_per_line) + "\n")
+        f.write(" " + str(field_width) + ", " + str(no_data) + ",  , 1 , 1" + "\n")
+        f.write("   " + str(ny) + ",  " + str(nx) + ",  " + str(xllcenter) + ",  " + str(hix) + ",  " + str(
+            yllcenter) + ",  " + str(hiy) + "\n")
+        f.write("   " + str(dx) + ",  0.0,  0.0    " + "\n")
+        f.write("@" + "\n")
+
+        for n in range(z.shape[1]):
+            count = 0
+            for m in range(z.shape[0] - 1, -1, -1):
+                count += 1
+                if np.isnan(z[m, n]):
+                    f.write(space_back_to_front(format(no_data, "13.7E") + '  '))
+                else:
+                    if abs(z[m, n]) >= 1E100:
+                        f.write(space_back_to_front(format(z[m, n], "13.7E") + ' '))
+                    elif abs(z[m, n]) >= 1E6:

```

Comment:
Where do these tolerance values come from?

Reply:
To correct for different string length due to extra exponential digit.